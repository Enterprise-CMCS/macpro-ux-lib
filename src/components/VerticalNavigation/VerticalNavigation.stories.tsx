import React from "react";
import { ComponentStory, ComponentMeta } from "@storybook/react";
import { VerticalNavigation } from "./VerticalNavigation";
export default {
  title: "COMPONENTS/Vertical Navigation",
  component: VerticalNavigation,

  argTypes: {
    items: {
      description:
        "Navigation section items to be rendered for the sidebar, can be up to three layers deep. \n\nExample object structure: `{ id: '1', items: [],  text: 'Parent'}`",
      control: false,
    },
    selectedId: {
      description:
        "An ID(s) of a section(s) on which should be programatically marked as currently active",
    },
  },
  args: {},
  parameters: {
    docs: {
      description: {
        component:
          "Hierarchical, vertical navigation to place at the side of a page. The recommendation is to display a navigational hierarchy with one to three levels. In theory it can be infinitely nested, but three layers max is recommended for a better user experience.",
      },
    },
  },
} as ComponentMeta<typeof VerticalNavigation>;

const Template: ComponentStory<typeof VerticalNavigation> = ({ ...rest }) => (
  <VerticalNavigation {...rest} />
);

export const Default = Template.bind({});
Default.args = {
  items: [
    { id: "1", items: [], text: "Parent" },
    {
      id: "2",
      items: [
        {
          id: "4",
          items: [
            { id: "6", text: "Grandchild" },
            { id: "7", text: "Grandchild" },
          ],
          text: "Child",
        },
        { id: "5", items: [], text: "Child" },
      ],

      text: "Parent",
    },
    { id: "3", items: [], text: "Parent" },
  ],
};

export const Togglable = Template.bind({});

Togglable.parameters = {
  docs: {
    description: {
      story:
        "Items with navigation children are given the option to be toggleable to show/hide their navigation options.",
    },
  },
};

Togglable.args = {
  items: [
    { id: "1", items: [], text: "Parent" },
    {
      id: "2",
      togglable: true,
      items: [
        {
          id: "4",
          togglable: true,
          items: [
            { id: "6", text: "Grandchild" },
            { id: "7", text: "Grandchild" },
          ],
          text: "Child with toggle",
        },
        { id: "5", items: [], text: "Child" },
      ],

      text: "Parent with toggle",
    },
    { id: "3", items: [], text: "Parent" },
  ],
};

export const SelectedId = Template.bind({});

SelectedId.parameters = {
  docs: {
    description: {
      story:
        "We can set a selected ID of one of the sections on the page and it will be marked on page load.",
    },
  },
};

SelectedId.args = {
  selectedId: "7",
  items: [
    { id: "1", items: [], text: "Parent" },
    {
      id: "2",
      togglable: true,
      items: [
        {
          id: "4",
          togglable: true,
          items: [
            { id: "6", text: "Grandchild" },
            { id: "7", text: "Grandchild" },
          ],
          text: "Child with toggle",
        },
        { id: "5", items: [], text: "Child" },
      ],

      text: "Parent with toggle",
    },
    { id: "3", items: [], text: "Parent" },
  ],
};
