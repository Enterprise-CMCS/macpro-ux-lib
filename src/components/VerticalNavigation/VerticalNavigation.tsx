import React, { useState } from "react";
import {
  NavigationItemChild,
  VerticalNavigationItem,
} from "./VerticalNavigationItem";
type IntrinsicElements = JSX.IntrinsicElements["nav"];

interface Props extends IntrinsicElements {
  items: IVerticalNavigationItem[];
  selectedId?: string;
}

export interface IVerticalNavigationItem {
  id: string;
  items?: NavigationItemChild[];
  togglable?: boolean;
  href?: string;
  text: string;
}

/**
 * Vertical Navigation Component
 * @param {IVerticalNavigationItem[]}              items            Navigation section items to be rendered for the sidebar, can be up to three layers deep.
 * @param {string}                        selectedId       Selected id of section that is currently active.
 */

export const VerticalNavigation: React.FC<Props> = ({
  selectedId,
  items,
  ...rest
}) => {
  const getSelectedSecitons = (): string[] => {
    const foundSelectedIds: string[] = [];
    items.some((parentItem) => {
      if (parentItem.id === selectedId) {
        foundSelectedIds.push(parentItem.id);
        return true;
      }

      parentItem.items?.some((childItem) => {
        if (childItem.id === selectedId) {
          foundSelectedIds.push(parentItem.id, childItem.id);
          return true;
        }

        childItem.items?.some((grandchildItem) => {
          if (grandchildItem.id === selectedId) {
            foundSelectedIds.push(
              parentItem.id,
              childItem.id,
              grandchildItem.id
            );
            return true;
          }
        });
      });
    });

    return foundSelectedIds;
  };

  const [selectedIds, setSelectedIds] = useState<string[]>(
    selectedId ? getSelectedSecitons() : []
  );

  return (
    <div className="tablet:grid-col-4 pointer">
      <nav aria-label="Side navigation" {...rest}>
        <ul className="usa-sidenav">
          {items.map((item) => {
            return (
              <VerticalNavigationItem
                selectedIds={selectedIds}
                navClick={setSelectedIds}
                id={item.id}
                text={item.text}
                items={item.items}
                key={item.id}
                togglable={item.togglable}
              />
            );
          })}
        </ul>
      </nav>
    </div>
  );
};
