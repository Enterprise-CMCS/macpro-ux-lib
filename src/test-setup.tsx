import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };
