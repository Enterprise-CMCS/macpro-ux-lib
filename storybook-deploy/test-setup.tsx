import React, { ReactElement } from "react";
import { render, RenderOptions } from "@testing-library/react";
import "@testing-library/jest-dom";

// Here we can set all providers for themes, stores and other stuff
const AllTheProviders = ({ children }: { children: JSX.Element }) => {
  return <>{children}</>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, "wrapper">
) => render(ui, { wrapper: AllTheProviders, ...options });

export * from "@testing-library/react";
export { customRender as render };

/**
 * Clean up attributes from the rendered component by removing the generated ids.
 *
 * Before:
 *   for="input-type-text-123456"
 *   id="input-type-text-123456"
 *
 * After:
 *   for="input-type-text"
 *   id="input-type-text"
 *
 * This is useful during snapshot testing. Snapshot comparisons will fail without addressing these items.
 *
 * @param {HTMLElement} container The component to clean up.
 * @param {prefixes}      string[]  The reference substrings to leave alone.
 */
