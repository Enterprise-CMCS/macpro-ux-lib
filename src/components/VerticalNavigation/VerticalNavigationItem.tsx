import { Icon } from "../Icon/Icon";
import React, { Dispatch, SetStateAction } from "react";
import { IVerticalNavigationItem } from "./VerticalNavigation";

export interface NavigationItemChild extends IVerticalNavigationItem {
  navClick: Dispatch<SetStateAction<string[]>>;
  selectedIds: string[];
}

/**
 * Vertical Navigation Child Component
 * @param {Dispatch<SetStateAction<string[]>>}     navClick       Click hanlder passed from the Veritical Navigation component that handles showing/hiding sections.
 * @param {string[]}                               selectedIds    An array of currently selected ids that is being tracked by parent component.
 */

export const VerticalNavigationItem: React.FC<NavigationItemChild> = ({
  navClick,
  id,
  selectedIds,
  text,
  togglable,
  items,
  href,
}) => {
  const generateSections = (
    id: string,
    text: string,
    togglable?: boolean,
    href?: string,
    items?: NavigationItemChild["items"],
    previousItemsIds: string[] = []
  ) => {
    return (
      <li className="usa-sidenav__item" key={id}>
        <a
          href={href}
          className={selectedIds.includes(id) ? "usa-current" : ""}
          onClick={() =>
            !selectedIds.includes(id)
              ? navClick([...previousItemsIds, id])
              : togglable
              ? navClick([
                  ...previousItemsIds.filter((prevId) => prevId !== id),
                ])
              : undefined
          }
        >
          {text}
          {togglable && (
            <span className="float-right">
              {selectedIds.includes(id) ? (
                <Icon name="arrow_drop_up" />
              ) : (
                <Icon name="arrow_drop_down" />
              )}
            </span>
          )}
        </a>
        {selectedIds.includes(id) && items && items.length > 0 && (
          <ul className="usa-sidenav__sublist">
            {items.map((itemChild) => {
              previousItemsIds.push(id);

              return generateSections(
                itemChild.id,
                itemChild.text,
                itemChild.togglable,
                itemChild.href,
                itemChild.items,
                previousItemsIds
              );
            })}
          </ul>
        )}
      </li>
    );
  };

  return <>{generateSections(id, text, togglable, href, items)}</>;
};
