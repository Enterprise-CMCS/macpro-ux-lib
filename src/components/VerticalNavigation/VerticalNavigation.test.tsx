import React from "react";
import { fireEvent, screen, render } from "../../test-setup";
import { VerticalNavigation } from "./VerticalNavigation";

describe("Tests for the VerticalNavigation component.", () => {
  const setState = jest.fn();
  const selectedIds: string[] = [];

  it("Should render out full sidebar nav component with options, an active section, and togglable", () => {
    render(
      <VerticalNavigation
        items={[
          {
            id: "2",
            togglable: true,
            items: [
              {
                id: "4",
                togglable: true,
                items: [
                  {
                    id: "6",
                    navClick: setState,
                    selectedIds,
                    text: "Grandchild",
                  },
                ],
                text: "Child",
                navClick: setState,
                selectedIds,
              },
            ],
            text: "Parent",
          },
        ]}
        data-testid="verticalNavigation"
        selectedId="6"
      />
    );
    const SideNavComp = screen.getByTestId("verticalNavigation");
    const parentSection = screen.getByText("Parent");
    const childSection = screen.getByText("Child");
    const grandChildSection = screen.getByText("Grandchild");
    fireEvent.click(grandChildSection);
    fireEvent.click(childSection);
    fireEvent.click(parentSection);

    expect(SideNavComp).toBeInTheDocument();
  });

  it("Should render out full sidebar nav component with options, an active section, and togglable", () => {
    const setState2 = jest.fn();
    render(
      <VerticalNavigation
        items={[
          {
            id: "2",
            togglable: true,
            items: [
              {
                id: "4",
                items: [
                  {
                    id: "6",
                    navClick: setState,
                    selectedIds: ["6", "4", "2"],
                    text: "Grandchild",
                  },
                ],
                text: "Child",
                navClick: setState,
                selectedIds: ["6", "4", "2"],
              },
            ],
            text: "Parent",
          },
        ]}
        data-testid="verticalNavigation"
        selectedId="6"
      />
    );

    const parentSection = screen.getByText("Parent");
    const childSection = screen.getByText("Child");
    const grandChildSection = screen.getByText("Grandchild");
    console.log(parentSection);
    fireEvent.click(grandChildSection);
    fireEvent.click(childSection);
    fireEvent.click(parentSection);
  });

  it("Should render out full sidebar nav component with options and an active section", () => {
    render(
      <VerticalNavigation
        items={[
          { id: "1", items: [], text: "Parent" },
          { id: "2", items: [], text: "Parent" },
        ]}
        data-testid="verticalNavigation"
        selectedId="1"
      />
    );

    const SideNavComp = screen.getByTestId("verticalNavigation");
    expect(SideNavComp).toBeInTheDocument();
  });

  it("Should render a vertical nav component with dropdown toggles", () => {
    render(
      <VerticalNavigation
        items={[
          { id: "1", items: [], text: "Parent" },
          {
            id: "2",
            togglable: true,
            items: [
              {
                id: "4",
                togglable: true,
                items: [
                  {
                    id: "6",
                    navClick: setState,
                    selectedIds,
                    text: "Grandchild",
                  },
                ],
                text: "Child",
                navClick: setState,
                selectedIds,
              },
              {
                id: "5",
                items: [],
                text: "Child",
                navClick: setState,
                selectedIds: selectedIds,
              },
            ],

            text: "Parent",
          },
          { id: "3", items: [], text: "Parent" },
        ]}
        selectedId="6"
        data-testid="verticalNavigation"
      />
    );

    const SideNavComp = screen.getByTestId("verticalNavigation");
    expect(SideNavComp).toHaveTextContent("Grandchild");
    expect(SideNavComp.getElementsByTagName("span")[0]).toBeInTheDocument();
  });

  it("Should render out full sidebar nav component with options and an active section in child", () => {
    render(
      <VerticalNavigation
        items={[
          {
            id: "1",
            items: [
              {
                id: "2",
                navClick: setState,
                selectedIds,
                items: [],
                text: "Parent",
              },
            ],
            text: "Parent",
          },
          { id: "3", items: [], text: "Parent" },
        ]}
        data-testid="verticalNavigation"
        selectedId="2"
      />
    );

    const SideNavComp = screen.getByTestId("verticalNavigation");
    expect(SideNavComp).toBeInTheDocument();
  });
});

describe("compontent snapshots", () => {
  const setState = jest.fn();
  const selectedIds: string[] = [];

  it("Basic VerticalNavigation", () => {
    const { container } = render(
      <VerticalNavigation
        items={[
          { id: "1", items: [], text: "Parent" },
          {
            id: "2",
            items: [
              {
                id: "4",
                items: [
                  {
                    id: "6",
                    navClick: setState,
                    selectedIds,
                    text: "Grandchild",
                  },
                  {
                    id: "7",
                    navClick: setState,
                    selectedIds,
                    text: "Grandchild",
                  },
                ],
                text: "Child",
                navClick: setState,
                selectedIds,
              },
              {
                id: "5",
                items: [],
                text: "Child",
                navClick: setState,
                selectedIds: selectedIds,
              },
            ],

            text: "Parent",
          },
          { id: "3", items: [], text: "Parent" },
        ]}
        data-testid="verticalNavigation"
      />
    );
    expect(container).toMatchSnapshot();
  });

  it("Alternative VerticalNavigation", () => {
    const { container } = render(
      <VerticalNavigation items={[]} data-testid="verticalNavigation" />
    );
    expect(container).toMatchSnapshot();
  });
});
