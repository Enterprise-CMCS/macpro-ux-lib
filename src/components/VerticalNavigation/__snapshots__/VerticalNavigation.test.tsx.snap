// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compontent snapshots Alternative VerticalNavigation 1`] = `
<div>
  <div
    class="tablet:grid-col-4 pointer"
  >
    <nav
      aria-label="Side navigation"
      data-testid="verticalNavigation"
    >
      <ul
        class="usa-sidenav"
      />
    </nav>
  </div>
</div>
`;

exports[`compontent snapshots Basic VerticalNavigation 1`] = `
<div>
  <div
    class="tablet:grid-col-4 pointer"
  >
    <nav
      aria-label="Side navigation"
      data-testid="verticalNavigation"
    >
      <ul
        class="usa-sidenav"
      >
        <li
          class="usa-sidenav__item"
        >
          <a
            class=""
          >
            Parent
          </a>
        </li>
        <li
          class="usa-sidenav__item"
        >
          <a
            class=""
          >
            Parent
          </a>
        </li>
        <li
          class="usa-sidenav__item"
        >
          <a
            class=""
          >
            Parent
          </a>
        </li>
      </ul>
    </nav>
  </div>
</div>
`;
